### PR TITLE
WIP Feature/master/tdi 37475 azure storage

### PR DESCRIPTION
**Pull request by undx**
**Please check if the PR fulfills these requirements**

- [ ] The commit message follows Talend standard
- [X] Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/928/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/928/)

**Comment by sgandon** [comment]
Could you move the component folder (*components-azurestorage*) into the subfolder named *components*. All of the others shall be moved there as well soon.

**Comment by sgandon** [comment]
I have added Spring and OSGI generic tests to make sure the components are properly registered and I18Ned

**Comment by sgandon** [comment]
could you explain to me how to launch the Integration tests please ?

**Comment by undx** [comment]
Thanks for the remarks.
To launch IT tests, with maven pass this parameters/variables : 
* `azurestorage.account.name`  : the storage account name.
* `azurestorage.account.key` : one of the two keys generated for the storage account.
* `azurestorage.sharedaccesssignature` : The generated SAS created for the storage.
* `azurestorage.usesas` : true or false. If true it uses  `azurestorage.sharedaccesssignature` otherwise it uses account informations.

I can send you by email my testing account credentials if you've no Azure storage account.

File: components-azurestorage/src/main/java/org/talend/components/azurestorage/runtime/AzureStorageContainerCreateReader.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+                containerPermissions.setPublicAccess(BlobContainerPublicAccessType.CONTAINER);
+                // Set the permissions on the container.
+                container.uploadPermissions(containerPermissions);
+            }
+            if (!result) {
+                LOGGER.warn("Container " + mycontainer + " already exists !");
+            }
+            dataCount++;
+            return result;
+        } catch (Exception e) {
```
Comment (undx) [inline components-azurestorage/src/main/java/org/talend/components/azurestorage/runtime/AzureStorageContainerCreateReader.java]: Ok.

File: components-azurestorage/src/main/java/org/talend/components/azurestorage/runtime/AzureStorageContainerCreateReader.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            LOGGER.error(e.getLocalizedMessage());
+            if (properties.dieOnError.getValue())
+                throw new ComponentException(e);
+            return result;
```
Comment (undx) [inline components-azurestorage/src/main/java/org/talend/components/azurestorage/runtime/AzureStorageContainerCreateReader.java]: Sorry, I didn't catch this one.

**Comment by undx** [comment]
I met a problem when running the tAzureStorageContainerList (tAzureStoragList
also) in the Studio which I hadn't before.
In [TAzureStorageContainerListProperties](https://github.com/Talend/components/blob/feature/master/TDI-37475_AzureStorage/components-azurestorage/src/main/java/org/talend/components/azurestorage/tazurestoragecontainerlist/TAzureStorageContainerListProperties.java#L46-L48), I defined a simple schema, my *[Reader](https://github.com/Talend/components/blob/feature/master/TDI-37475_AzureStorage/components-azurestorage/src/main/java/org/talend/components/azurestorage/runtime/AzureStorageContainerListReader.java#L31) is quit simple too.
The job fails because I've different schemas :
* Design  schema is `{"type":"record","name":"Main","fields":[{"name":"ContainerName","type":"string","talend.field.length":"50"}]}`
* Runtime schema is
`{"type":"record","name":"StringRecord","fields":[{"name":"field","type":"string"}]}`

This leads to a NPE [here](https://github.com/Talend/daikon/blob/master/daikon/src/main/java/org/talend/daikon/di/IndexMapperByName.java#L53-L54) due to the fact that the design schema field isn't present in the runtime schema.
Am I missing something or ... ?

**Comment by gonchar-ivan** [comment]
Design schema and runtime schema should be the same in your case. 
I wonder why they are different

**Comment by gonchar-ivan** [comment]
You need to fix it in following way:
Adjust your Reader to return IndexedRecord rather than String from getCurrent() method.
1. Any Reader should have IndexedRecord currentRecord field. 
(Multiple calls to getCurrent() should return the same result until advance is called. Check Reader interface javadocs.)
2. You should create a new instance of IndexedRecord in Reader.advance() method. I propose you to do:

```java
IndexedRecord record = new GenericData.Record(schema);
record.put(0, "stringFromSource");
```
where schema is equal to design schema 
(you need to retrieve it from properties instance)

getCurrent() should be:

```java
    @Override
    public String getCurrent() throws NoSuchElementException {
        return currentRecord;
    }
```

**Comment by undx** [comment]
Thanks @gonchar-ivan for the advices!

That's what I've planned to do to bypass the problem (convert to IndexedRecord). 
As the component is returning only one single String, I followed this [section](https://github.com/Talend/components/wiki/6.-Component-Runtime#emitting-primitives) in the wiki (Emitting primitives).
But the strange thing is that it was working before.
Thanks anyway.

**Comment by sgandon** [comment]
Hello guy, I must apologize cause I did push a full maven repo on this branch.
I will fix this but I'll have to do a push force to remove this from the history and avoid increase the size of the repo for nothing.
So please do not commit anything until I fixed it.
Sorry again.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/1172/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/1172/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/172/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/172/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/264/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/264/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/289/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/289/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/513/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/513/)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/523/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/523/)

**Pull request closed without merging**